### PR TITLE
Added automatic entitlement certificate regeneration on product/content update

### PR DESCRIPTION
**Pull request by Ceiu**
- Entitlement certificates will now be regenerated when an associated
  products or content are updated or removed
- Refactored much of the entitlement certificate regeneration code. A
  new class, EntitlementCertificateGenerator, now houses most of the
  regeneration bits that were previously in CandlepinPoolManager.
- Moved the product and content CRUD operations from the curators
  to the new ProductManager and ContentManager classes
- Several other stylistic and checkstyle related code changes
- Added tests to improve or complete test coverage on several
  methods which were previously overlooked

### CODE REVIEW

**Comment by Ceiu** [comment]
Those of you who have recently had commits merged into master, please take a close look at those bits in this branch to make sure they didn't get clobbered while resolving merge conflicts.

**Comment by wottop** [comment]
buildbot: retest this please

**Comment by Ceiu** [comment]
Looks like one of the methods I used in my test changed slightly on me. Looking into it and will rebase/update as necessary.

**Comment by Ceiu** [comment]
@candlepin/jenkins-bot retest this please

File: server/src/main/java/org/candlepin/resource/OwnerContentResource.java [hunk 1]
```diff
@@ -279,11 +271,6 @@ public void remove(@PathParam("owner_key") String ownerKey,
             throw new ForbiddenException(i18n.tr("content \"{1}\" is locked", content.getId()));
         }
 
-        List<Product> affectedProducts = this.contentCurator.removeContent(content, owner);
-
-        // Regenerate affected product certs as dirty...
-        for (Product product : affectedProducts) {
-            poolManager.regenerateCertificatesOf(owner, product.getId(), true);
-        }
+        List<Product> affectedProducts = this.contentManager.removeContent(content, owner, true);
```
Comment (nguyenfilip) [inline server/src/main/java/org/candlepin/resource/OwnerContentResource.java]: unused collection affectedProducts

**Comment by nguyenfilip** [comment]
@Ceiu Why cp2_owner_content and cp2_owner_products doesn't have foreign key constraints?

File: server/src/main/java/org/candlepin/controller/EntitlementCertificateGenerator.java [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+
+/**
+ * The EntitlementCertificateGenerator class provides utility methods for regenerating the
+ * certificates for entitlements associated with various model objects.
+ * <p/>
+ * It should be noted that due to the amount of graph traversal that can occur when performing these
+ * operations, that the number of calls to methods provided by this class should be minimized. The
+ * majority are incredibly expensive and, in the case of immediate regeneration, can hold database
+ * locks for the duration. Usage of these methods should be carefully evaluated, and bulk operations
+ * should be prefered to singular ones.
```
Comment (nguyenfilip) [inline server/src/main/java/org/candlepin/controller/EntitlementCertificateGenerator.java]: Typo "preferred"

File: server/src/main/java/org/candlepin/controller/EntitlementCertificateGenerator.java [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+import java.util.Set;
+
+
+
+/**
+ * The EntitlementCertificateGenerator class provides utility methods for regenerating the
+ * certificates for entitlements associated with various model objects.
+ * <p/>
+ * It should be noted that due to the amount of graph traversal that can occur when performing these
+ * operations, that the number of calls to methods provided by this class should be minimized. The
```
Comment (nguyenfilip) [inline server/src/main/java/org/candlepin/controller/EntitlementCertificateGenerator.java]: Typo "these operations, that the" -> "these operations, the"

File: server/src/main/java/org/candlepin/controller/ContentManager.java [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+                alt.addOwner(owner);
+                this.contentCurator.merge(alt);
+
+                // Make sure every product using the old version/entity are updated to use the new one
+                List<Product> affectedProducts = this.productCurator.getProductsWithContent(
+                    owner, Arrays.asList(entity.getId())
+                );
+
+                for (Product product : affectedProducts) {
+                    product = (Product) product.clone();
```
Comment (nguyenfilip) [inline server/src/main/java/org/candlepin/controller/ContentManager.java]: I am not completely sure but it seems to me that the part of the update content that modifies affectedProducts is not covered by any spec test - no spec test has product with content that is being converged.

Comment (Ceiu) [inline server/src/main/java/org/candlepin/controller/ContentManager.java]: product_versioning_spec.rb => "creates new products when adding content to shared products" and "creates new products when updating content used by shared products"

Comment (nguyenfilip) [inline server/src/main/java/org/candlepin/controller/ContentManager.java]: I am talking about the code block starting at line 187 - the for loop

Comment (Ceiu) [inline server/src/main/java/org/candlepin/controller/ContentManager.java]: Added two tests to cover this

**Comment by Ceiu** [comment]
Gross oversight on the lack of FKs on the owner_products/content tables.

**Comment by Ceiu** [comment]
@nguyenfilip All the requested changes, as well as several others from @vritant are in.

**Comment by nguyenfilip** [comment]
@Ceiu Ack, will merge this after checkstyle passes. I think you will have to rebase against master to make it pass

**Pull request merged**